pkg/listener/job_processor.go

```diff
@@ -205,15 +207,27 @@ func (p *JobProcessor) getJobWithRetries(jobID string) (*api.JobRequest, error)
 }
 
 func (p *JobProcessor) StopJob(jobID string) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	// The job finished before the sync request returned a stop-job command.
+	// Here, we don't do anything since the job is already finished and
+	// a finished-job state will be reported in the next sync.
+	if p.State == selfhostedapi.AgentStateFinishedJob {
+		return
+	}
+
 	p.CurrentJobID = jobID
 	p.State = selfhostedapi.AgentStateStoppingJob
 
 	p.CurrentJob.Stop()
 }
 
 func (p *JobProcessor) JobFinished(result selfhostedapi.JobResult) {
+	p.mutex.Lock()
```

Comment:
why not defer here as well?

Reply:
This method has no extra branches, like the other one.